File: src/main/java/com/joecollins/graphics/components/lowerthird/LowerThirdHeadlineAndSummaryBothEnds.java
package com.joecollins.graphics.components.lowerthird;

import com.joecollins.bindings.Binding;
import com.joecollins.graphics.components.lowerthird.SummaryFromBothEnds.Entry;
import java.awt.BorderLayout;
import java.awt.GridLayout;
import javax.swing.JPanel;

public class LowerThirdHeadlineAndSummaryBothEnds extends LowerThird {

  private HeadlinePanel headlinePanel = new HeadlinePanel();
  private SummaryFromBothEnds partySummary = new SummaryFromBothEnds();

  private Binding<String> headlineBinding = Binding.fixedBinding("");
  private Binding<String> subheadBinding = Binding.fixedBinding(null);

  public LowerThirdHeadlineAndSummaryBothEnds() {
    JPanel center = new JPanel();
    center.setLayout(new GridLayout(1, 2));
    add(center, BorderLayout.CENTER);

    center.add(headlinePanel);
    center.add(partySummary);
  }

  String getHeadline() {
    return headlinePanel.getHeadline();
  }

  public void setHeadlineBinding(Binding<String> headlineBinding) {
    this.headlineBinding.unbind();
    this.headlineBinding = headlineBinding;
    this.headlineBinding.bind(headlinePanel::setHeadline);
  }

  String getSubhead() {
    return headlinePanel.getSubhead();
  }

  public void setSubheadBinding(Binding<String> subheadBinding) {
    this.subheadBinding.unbind();
    this.subheadBinding = subheadBinding;
    this.subheadBinding.bind(headlinePanel::setSubhead);
  }

  String getSummaryHeader() {
    return this.partySummary.getHeadline();
  }

  public void setSummaryHeaderBinding(Binding<String> summaryHeaderBinding) {
    this.partySummary.setHeadlineBinding(summaryHeaderBinding);
  }

  public int getTotal() {
    return partySummary.getTotal();
  }

  public void setTotalBinding(Binding<Integer> totalBinding) {
    partySummary.setTotalBinding(totalBinding);
  }

  public Entry getLeft() {
    return partySummary.getLeft();
  }

  public void setLeftBinding(Binding<Entry> leftBinding) {
    partySummary.setLeftBinding(leftBinding);
  }

  public Entry getRight() {
    return partySummary.getRight();
  }

  public void setRightBinding(Binding<Entry> rightBinding) {
    partySummary.setRightBinding(rightBinding);
  }

  public Entry getMiddle() {
    return partySummary.getMiddle();
  }

  public void setMiddleBinding(Binding<Entry> middleBinding) {
    partySummary.setMiddleBinding(middleBinding);
  }
}

File: src/main/java/com/joecollins/graphics/components/lowerthird/LowerThirdHeadlineAndSummaryBothEnds.kt
package com.joecollins.graphics.components.lowerthird

import com.joecollins.bindings.Binding
import java.awt.BorderLayout
import java.awt.GridLayout
import java.awt.Image
import java.time.Clock
import java.time.ZoneId
import javax.swing.JPanel

class LowerThirdHeadlineAndSummaryBothEnds internal constructor(
    leftImageBinding: Binding<Image>,
    placeBinding: Binding<String>,
    timezoneBinding: Binding<ZoneId>,
    private val headlineBinding: Binding<String?>,
    private val subheadBinding: Binding<String?>,
    summaryHeaderBinding: Binding<String>,
    summaryTotalBinding: Binding<Int>,
    summaryLeftBinding: Binding<SummaryFromBothEnds.Entry?>,
    summaryRightBinding: Binding<SummaryFromBothEnds.Entry?>,
    summaryMiddleBinding: Binding<SummaryFromBothEnds.Entry?> = Binding.fixedBinding(null),
    clock: Clock,
    showTimeZone: Boolean = false
) : LowerThird(leftImageBinding, placeBinding, timezoneBinding, clock, showTimeZone) {

    constructor(
        leftImageBinding: Binding<Image>,
        placeBinding: Binding<String>,
        timezoneBinding: Binding<ZoneId>,
        headlineBinding: Binding<String?>,
        subheadBinding: Binding<String?>,
        summaryHeaderBinding: Binding<String>,
        summaryTotalBinding: Binding<Int>,
        summaryLeftBinding: Binding<SummaryFromBothEnds.Entry?>,
        summaryRightBinding: Binding<SummaryFromBothEnds.Entry?>,
        summaryMiddleBinding: Binding<SummaryFromBothEnds.Entry?> = Binding.fixedBinding(null),
        showTimeZone: Boolean = false
    ) : this(
        leftImageBinding,
        placeBinding,
        timezoneBinding,
        headlineBinding,
        subheadBinding,
        summaryHeaderBinding,
        summaryTotalBinding,
        summaryLeftBinding,
        summaryRightBinding,
        summaryMiddleBinding,
        Clock.systemDefaultZone(),
        showTimeZone
    )

    private val headlinePanel = HeadlinePanel()
    private val partySummary = SummaryFromBothEnds(summaryHeaderBinding, summaryTotalBinding, summaryLeftBinding, summaryRightBinding, summaryMiddleBinding)

    internal val headline: String?
        get() = headlinePanel.headline

    internal val subhead: String?
        get() = headlinePanel.subhead

    internal val summaryHeader: String
        get() = partySummary.headline

    internal val total: Int
        get() = partySummary.total

    internal val left: SummaryFromBothEnds.Entry?
        get() = partySummary.left

    internal val right: SummaryFromBothEnds.Entry?
        get() = partySummary.right

    internal val middle: SummaryFromBothEnds.Entry?
        get() = partySummary.middle

    init {
        val center = JPanel()
        center.layout = GridLayout(1, 2)
        add(center, BorderLayout.CENTER)
        center.add(headlinePanel)
        center.add(partySummary)
        this.headlineBinding.bind { headlinePanel.headline = it }
        this.subheadBinding.bind { headlinePanel.subhead = it }
    }
}
